### PR TITLE
[mqtt] Add extended device info

### PR DESCRIPTION
**Pull request by dentra**
# What does this implement/fix?

This PR set model and manufacturer by following esphome.project.name rule and set hw_version if esphome.project.version defined. Additionally adds MAC address to device connections.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/mqtt/mqtt_component.cpp [hunk 1]
```diff
@@ -150,12 +150,29 @@ bool MQTTComponent::send_discovery_() {
...
+          device_info[MQTT_DEVICE_MODEL] = ESPHOME_BOARD;
+          device_info[MQTT_DEVICE_MANUFACTURER] = ESPHOME_PROJECT_NAME;
+        }
+#else
         device_info[MQTT_DEVICE_MODEL] = ESPHOME_BOARD;
         device_info[MQTT_DEVICE_MANUFACTURER] = "espressif";
+#endif
         device_info[MQTT_DEVICE_SUGGESTED_AREA] = node_area;
+#ifdef ESPHOME_PROJECT_VERSION
+        device_info[MQTT_DEVICE_HW_VERSION] = ESPHOME_PROJECT_VERSION;
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_component.cpp]: I am changing this on the HA side to make the project version the main one.

https://github.com/home-assistant/core/pull/123183

Comment (dentra) [reply esphome/components/mqtt/mqtt_component.cpp]: Your change affects only API, but not MQTT.
A this moment sw_version is “esphome v2024.7.3 6 Aug 00:25:23”. Is it ok if I’ll sync it with your changes to “1.2.3 (ESPHome 2024.7.3)” when project.version is defined and “2024.7.3 (6 Aug 00:25:23) when not?

Comment (jesserockz) [reply esphome/components/mqtt/mqtt_component.cpp]: That is what I was hinting at :wink:

File: esphome/components/mqtt/mqtt_component.cpp [hunk 2]
```diff
@@ -150,12 +150,30 @@ bool MQTTComponent::send_discovery_() {
...
+          device_info[MQTT_DEVICE_MODEL] = ESPHOME_BOARD;
+          device_info[MQTT_DEVICE_MANUFACTURER] = ESPHOME_PROJECT_NAME;
+        } else {
+          device_info[MQTT_DEVICE_MODEL] = model + 1;
+          device_info[MQTT_DEVICE_MANUFACTURER] = std::string(ESPHOME_PROJECT_NAME, model - ESPHOME_PROJECT_NAME);
+        }
+#else
+        device_info[MQTT_DEVICE_SW_VERSION] = ESPHOME_VERSION " (" + App.get_compilation_time() + ")";
         device_info[MQTT_DEVICE_MODEL] = ESPHOME_BOARD;
         device_info[MQTT_DEVICE_MANUFACTURER] = "espressif";
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_component.cpp]:
```suggestion
#if defined(USE_ESP8266) || defined(USE_ESP32)
		    device_info[MQTT_DEVICE_MANUFACTURER] = "Espressif";
#elif defined(USE_RP2040)
		    device_info[MQTT_DEVICE_MANUFACTURER] = "Raspberry Pi";
#elif defined(USE_BK72XX)
		    device_info[MQTT_DEVICE_MANUFACTURER] = "Beken";
#elif defined(USE_RTL87XX)
		    device_info[MQTT_DEVICE_MANUFACTURER] = "Realtek";
#elif defined(USE_HOST)
		    device_info[MQTT_DEVICE_MANUFACTURER] = "Host";
#endif
```

**Review by jesserockz — DISMISSED** [review]

File: esphome/components/mqtt/mqtt_component.cpp [hunk 3]
```diff
@@ -150,12 +150,40 @@ bool MQTTComponent::send_discovery_() {
...
+#if defined(USE_ESP8266) || defined(USE_ESP32)
+		    device_info[MQTT_DEVICE_MANUFACTURER] = "Espressif";
+#elif defined(USE_RP2040)
+		    device_info[MQTT_DEVICE_MANUFACTURER] = "Raspberry Pi";
+#elif defined(USE_BK72XX)
+		    device_info[MQTT_DEVICE_MANUFACTURER] = "Beken";
+#elif defined(USE_RTL87XX)
+		    device_info[MQTT_DEVICE_MANUFACTURER] = "Realtek";
+#elif defined(USE_HOST)
+		    device_info[MQTT_DEVICE_MANUFACTURER] = "Host";
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_component.cpp]:
```suggestion
#if defined(USE_ESP8266) || defined(USE_ESP32)
        device_info[MQTT_DEVICE_MANUFACTURER] = "Espressif";
#elif defined(USE_RP2040)
        device_info[MQTT_DEVICE_MANUFACTURER] = "Raspberry Pi";
#elif defined(USE_BK72XX)
        device_info[MQTT_DEVICE_MANUFACTURER] = "Beken";
#elif defined(USE_RTL87XX)
        device_info[MQTT_DEVICE_MANUFACTURER] = "Realtek";
#elif defined(USE_HOST)
        device_info[MQTT_DEVICE_MANUFACTURER] = "Host";
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**